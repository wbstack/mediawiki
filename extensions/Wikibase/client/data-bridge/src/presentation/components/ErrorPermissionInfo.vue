<template>
	<div class="wb-db-error-permission-info">
		<div
			class="wb-db-error-permission-info__header"
			v-html="messageHeader"
		/>
		<a
			class="wb-db-error-permission-info__toggle"
			:class="[ `wb-db-error-permission-info__toggle--${ infoIsExpanded ? 'open' : 'closed' }` ]"
			@click="toggleInfo"
		>
			{{ this.$messages.getText( this.$messages.KEYS.PERMISSIONS_MORE_INFO ) }}
		</a>
		<div
			class="wb-db-error-permission-info__body"
			v-html="messageBody"
			v-if="infoIsExpanded"
		/>
	</div>
</template>

<script lang="ts">
import {
	Prop,
	Vue,
} from 'vue-property-decorator';
import Component from 'vue-class-component';

/**
 * A component used to illustrate permission errors which happened when
 * checking the user's authorization to perform an action.
 */
@Component
export default class ErrorPermissionInfo extends Vue {
	public infoIsExpanded = false;

	/**
	 * Flag to decide if the component is to be shown in an expanded
	 * state initially.
	 */
	@Prop( { required: false, default: false, type: Boolean } )
	public expandedByDefault!: boolean;

	/**
	 * The mark-up to show in the header.
	 * Careful, this value will not be HTML-escaped for you to allow
	 * for formatting of the content.
	 */
	@Prop( { required: true } )
	private readonly messageHeader!: string;

	/**
	 * The mark-up to show in the body.
	 * Careful, this value will not be HTML-escaped for you to allow
	 * for formatting of the content.
	 */
	@Prop( { required: true } )
	private readonly messageBody!: string;

	public created(): void {
		this.infoIsExpanded = this.expandedByDefault;
	}

	private toggleInfo(): void {
		this.infoIsExpanded = !this.infoIsExpanded;
	}
}
</script>

<style lang="scss">
.wb-db-error-permission-info {

	// TODO could use a variant of IconMessageBox
	padding-left: 2em;
	background: $svg-info no-repeat top left;

	@include body-responsive();

	&__header,
	p,
	ul {
		margin-bottom: $base-spacing-unit;
	}

	li:not( :first-child ) {
		margin-top: $margin-top-li;
	}

	&__toggle {
		$background-size: 14px;
		display: block;
		margin: $base-spacing-unit 0 2*$base-spacing-unit;
		padding-left: $background-size + 4px;
		background-position: top left;
		background-repeat: no-repeat;
		background-size: $background-size $background-size;
	}

	a,
	&__toggle {
		color: $color-primary;
		cursor: pointer;
	}

	a:hover,
	&__toggle:hover {
		color: $color-primary--hover;
	}

	&__toggle--closed {
		background-image: $svg-expand;
	}

	&__toggle--open {
		background-image: $svg-collapse;
	}
}
</style>
