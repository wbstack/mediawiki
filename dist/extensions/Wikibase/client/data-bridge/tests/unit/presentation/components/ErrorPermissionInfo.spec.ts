import ErrorPermissionInfo from '@/presentation/components/ErrorPermissionInfo.vue';
import { shallowMount } from '@vue/test-utils';
import MessageKeys from '@/definitions/MessageKeys';

const TOGGLE_SELECTOR = '.wb-db-error-permission-info__toggle';
const BODY_SELECTOR = '.wb-db-error-permission-info__body';

describe( 'ErrorPermissionInfo', () => {
	it( 'matches the snapshot in closed state', () => {
		const messageHeader = 'header';
		const messageBody = 'body';
		const expandedByDefault = false;
		const wrapper = shallowMount( ErrorPermissionInfo, {
			propsData: { messageHeader, messageBody, expandedByDefault },
		} );
		expect( wrapper.element ).toMatchSnapshot();
	} );

	it( 'matches the snapshot in opened state', () => {
		const messageHeader = 'header';
		const messageBody = 'body';
		const expandedByDefault = true;
		const wrapper = shallowMount( ErrorPermissionInfo, {
			propsData: { messageHeader, messageBody, expandedByDefault },
		} );
		expect( wrapper.element ).toMatchSnapshot();
	} );

	it( 'uses correct message for toggle', () => {
		const toggleText = 'mytoggle';
		const messageGet = jest.fn().mockReturnValue( toggleText );
		const wrapper = shallowMount( ErrorPermissionInfo, {
			propsData: { messageHeader: 'header', messageBody: 'body' },
			global: {
				mocks: {
					$messages: {
						KEYS: MessageKeys,
						getText: messageGet,
					},
				},
			},
		} );
		expect( messageGet ).toHaveBeenCalledTimes( 1 );
		expect( messageGet ).toHaveBeenCalledWith( MessageKeys.PERMISSIONS_MORE_INFO );
		expect( wrapper.find( TOGGLE_SELECTOR ).text() )
			.toBe( toggleText );
	} );

	it( 'Link toggles body visibility and link class', async () => {
		const expandedByDefault = false;
		const wrapper = shallowMount( ErrorPermissionInfo, {
			propsData: { messageHeader: 'header', messageBody: 'body', expandedByDefault },
		} );

		expect( wrapper.find( BODY_SELECTOR ).exists() )
			.toBeFalsy();

		await wrapper.find( TOGGLE_SELECTOR ).trigger( 'click' );

		expect( wrapper.find( TOGGLE_SELECTOR ).classes() )
			.toContain( 'wb-db-error-permission-info__toggle--open' );
		expect( wrapper.find( BODY_SELECTOR ).exists() )
			.toBeTruthy();

		await wrapper.find( TOGGLE_SELECTOR ).trigger( 'click' );

		expect( wrapper.find( TOGGLE_SELECTOR ).classes() )
			.toContain( 'wb-db-error-permission-info__toggle--closed' );
		expect( wrapper.find( BODY_SELECTOR ).exists() )
			.toBeFalsy();
	} );

} );
