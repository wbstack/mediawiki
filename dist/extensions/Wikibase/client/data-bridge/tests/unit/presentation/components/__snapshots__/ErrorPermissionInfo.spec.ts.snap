// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorPermissionInfo matches the snapshot in closed state 1`] = `
<div
  class="wb-db-error-permission-info"
>
  <div
    class="wb-db-error-permission-info__header"
  >
    header
  </div>
  <a
    class="wb-db-error-permission-info__toggle wb-db-error-permission-info__toggle--closed"
  >
    ⧼wikibase-client-data-bridge-permissions-error-info⧽
  </a>
  <!--v-if-->
</div>
`;

exports[`ErrorPermissionInfo matches the snapshot in opened state 1`] = `
<div
  class="wb-db-error-permission-info"
>
  <div
    class="wb-db-error-permission-info__header"
  >
    header
  </div>
  <a
    class="wb-db-error-permission-info__toggle wb-db-error-permission-info__toggle--open"
  >
    ⧼wikibase-client-data-bridge-permissions-error-info⧽
  </a>
  <div
    class="wb-db-error-permission-info__body"
  >
    body
  </div>
</div>
`;
